### DISCUSSION TITLE
Include symbols in parsed markdown?

### DISCUSSION THREAD

**Discussion by ianwsperber** [remarkjs/remark#646] [Q&A] [answered]
Hi all! Only just started using the remark toolkit, but I'm trying to write a markdown text editor and am curious for what advice anyone could offer on how to preserve the markdown symbols (eg `#` for headings, `_` for italics) in the parsed text? I'm currently using `remark-parse` to parse the AST and am hoping I can just write an extension to accomplish this. Thanks!

**Comment by wooorm** [comment, marked as answer]
We provide ASTs, not CSTs. See also: https://github.com/syntax-tree/mdast/issues/36.

Maybe preserving the syntax is an [XY problem](https://xyproblem.info)?

You could use the positional info on nodes, with the source file contents, to access whether for example emphasis was made with `*` or `_`

**Reply by ianwsperber** [reply]
Very helpful to see the context from another question, thank you! I wasn’t aware of the distinction between an AST & a CST.

Preserving the syntax is not an xy problem (it is the exact feature I am trying to build) but certainly possible remark is not the way to solve it

**Reply by ChristianMurphy** [reply]
> Preserving the syntax is not an xy problem (it is the exact feature I am trying to build) but certainly possible remark is not the way to solve it

It may still be an XY problem.
There are many editors built on markdown + mdast + remark which work just fine without encoding the symbols in an AST/CST.
For example https://github.com/bytedance/bytemd, https://github.com/josestg/mde, and https://github.com/theodorechu/append-editor

The missing context on why the exact syntax is needed and why the syntax tree needs to encode it are important to effectively answering your question.

**Reply by ianwsperber** [reply]
Oh that’s great context, thanks Christian! I’ll look through the source of those editors to learn about the proper implementation, exactly the kind of resource I was hoping to be linked to. Totally hear you that it may not be necessary to encode symbols in the AST, I may have framed my original question incorrectly , when I said “parsed text” I meant the final formatted text rendered in the browser. So would like the end user to be able to view the symbols after it’s been formatted (which I think at least MDE may do, will need to look more closely at your links this evening)

**Reply by ChristianMurphy** [reply]
> I said “parsed text” I meant the final formatted text rendered in the browser. So would like the end user to be able to view the symbols after it’s been formatted

I'm a bit confused here, these two sentences describe two very different things.
One a markdown to HTML transform for the browser, the other a markdown to markdown transform, unless I'm misunderstanding?

For markdown to HTML, you probably would not need to markdown symbols, as they are intentionally replaced with their HTML equivalents.
For formatting markdown back to markdown, it is possible, https://prettier.io uses remark+mdast to do markdown source formatting, but that's usually independent of the markdown to HTML aspect.

**Comment by Saul-Mirone** [comment]
For anyone looking for a solution to keep the user marker of emphasis and strong:

```ts
export const keepSymbolPlugin = () => {
    function transformer(ast, file) {
        visit(ast, node => ['strong', 'emphasis'].includes(node.type), (node) => {
            const marker = file.value.charAt(node.position.start.offset)
            node.marker = marker;
        });
    }
    return transformer;
};

unified()
    .use(remarkParse)
    .use(keepSymbolPlugin)
    .use(remarkStringify, {
        handlers: {
            strong: (node, _, state, info) => {
                const marker = node.marker || state.options.strong || '*';
                const exit = state.enter('strong')
                const tracker = state.createTracker(info)
                let value = tracker.move(marker + marker)
                value += tracker.move(
                    state.containerPhrasing(node, {
                    before: value,
                    after: marker,
                    ...tracker.current()
                    })
                )
                value += tracker.move(marker + marker)
                exit()
                return value
            },
            emphasis: (node, _, state, info) => {
                const marker = node.marker || state.options.emphasis || '*';
                const exit = state.enter('emphasis')
                const tracker = state.createTracker(info)
                let value = tracker.move(marker)
                value += tracker.move(
                    state.containerPhrasing(node, {
                    before: value,
                    after: marker,
                    ...tracker.current()
                    })
                )
                value += tracker.move(marker)
                exit()
                return value
            }
        }
    })
```
